bitbake-setup: fix key-error when changing a new 'section'

js731ca opened the pull request:
Setting a key=value in a new section would raise a key-error when using in-memory settings, e.g:
```
./bin/bitbake-setup --setting default top-dir-prefix /tmp/bitbake-setup --setting default top-dir-name gs settings foo bar baz

Loading settings from
    /tmp/bitbake-setup/gs/settings.conf

Traceback (most recent call last):
  File "/tmp/bitbake-setup/bitbake/./bin/bitbake-setup", line 853, in <module>
    main()
  File "/tmp/bitbake-setup/bitbake/./bin/bitbake-setup", line 838, in main
    all_settings = merge_settings(builtin_settings, global_settings, topdir_settings, args.setting)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/bitbake-setup/bitbake/./bin/bitbake-setup", line 733, in merge_settings
    all_settings[section][setting] = value
    ~~~~~~~~~~~~^^^^^^^^^
KeyError: 'foo'
```

kanavin commented on bin/bitbake-setup:
```diff
@@ -728,6 +728,8 @@ def merge_settings(builtin_settings, global_settings, local_settings, cmdline_se
+                    all_settings[section] = {}
                 all_settings[section][setting] = value
 
     for (section, setting, value) in cmdline_settings:
```
does this loop (over commandline overrides) need the same fix?

I also wonder if rather than allow setting 'unknown' settings, they should be checked against an 'allowlist', e.g.
```
if section not in all_settings.keys() or setting not in all_settings[section]:
    raise Exception('Unknown setting ..., please check bitbake-setup settings --list for available settings that can be set')
```
This check would be performed when either trying to set something, or override it from command line.

js731ca replied:
> does this loop (over commandline overrides) need the same fix?

good point... looking into it

> I also wonder if rather than allow setting 'unknown' settings, they should be checked against an 'allowlist', e.g.

that we could do - depends on what you had in mind with future additions to the settings

kanavin replied:
I kinda make my mind up as we go along. :)

I don't think it should be possible to set something that is not in the builltin set, because that invites code like:

```
if some_setting in settings[section]:
    do_stuff_with_that_setting
else:
   do_different_stuff
```
E.g. there would be code paths for when setting is present and when it isn't, complicating the logic, and sprinkling implicit defaults all over the place. Rather it should always be:
```
do_stuff_with(settings[section][some_setting])
```

js731ca replied:
Ok - so the builtin-settings are the "gold standard" to compare against.

Raising an exception might be to harsh: if the standard changes (e.g. a setting was removed from the default) a user might get locked out when coming from an older version -> printing a warning and ignoring the then unknown key?

js731ca committed:
bitbake-setup: rename function 'default_settings_path' to 'topdir_settings_path'

Rename the function to align with the corresponding
'global_settings_path' function, and move it down just above the
later.

This is done to differentiate from the built-in default settings, and
to align with the other places in the code that use 'topdir_settings'
(or 'global_settings')

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca commented:
v2: reworks the code to instead check sections and settings against the builtin 'default_settings' and skipping/dropping unknown keys.
(all that effort over three configuration values... ;-D)

kanavin commented on bin/bitbake-setup:
```diff
@@ -18,9 +18,18 @@ import configparser
+default_settings['default'] = {
+                 'top-dir-prefix':os.path.expanduser('~'),
+                 'top-dir-name':'bitbake-builds',
+                 'registry':default_registry,
```
What happens if someone tries to set (or override) dl-dir? It isn't in here (because it's set dynamically from top-dir in the builtin set), and so they will get warnings?

I'd say there should be an allowlist here (e.g. sections and keys, but not values), and then any checks are performed against that.

js731ca replied:
or we give it already a default? see 6b784b6
the benefit i see: having a fully populated default as reference keeps us from having a default, and an allow list = two things to maintain, vs one

kanavin replied:
```
                 'dl-dir': os.path.join(os.path.expanduser('~'), 'bitbake-builds', , '.bitbake-setup-downloads'),
```

If someone changes top dir prefix or top dir name to something else in global settings (a common use case), this will no longer work. The default dl-dir needs to be deduced from those two.

Once again, please prioritize #17 at this point.

js731ca replied:
(PR17 got some priority - and PR18 emerged, which i actually wanted to tackle this WE :-)

With the new way settings are merged, and the default having the lowest priority - i think this would still work.
Also the tests agree, and they do set a global dl-dir :-)

kanavin replied:
> With the new way settings are merged, and the default having the lowest priority - i think this would still work. Also the tests agree, and they do set a global dl-dir :-)

But it still makes the code more confusing. The dl-dir value from default_settings is overwritten after top_dir is determined, and never actually used, but anyone reading the code is going to be baffled as to why and how, and why is it even there, and set the way it is.

I'd also say setting the builtin defaults statically with a global assignment to constants like this may become limiting in the future. It's a lot more flexible to determine them in a function, maybe for now only dl-dir benefits from it, but who knows how bitbake-setup and its set of settings is going to evolve.

kanavin commented on bin/bitbake-setup:
```diff
@@ -852,9 +852,6 @@ def main():
 
-        # This cannot be set with the rest of the builtin settings as top_dir needs to be determined first
-        builtin_settings['default']['dl-dir'] = os.path.join(top_dir, '.bitbake-setup-downloads')
-
```
I had missed this change when I wrote the previous comment. This breaks the following scenario:

- someone sets top-dir-prefix in a global settings to somewhere like /some/special/yocto/directory and expects that dl-dir will automatically follow that to be contained inside that special directory. 

With the current code, it will work, because dl-dir is programmatically deduced from top-dir. With this change, it no longer will. Tests do not expose the breakage because they set dl-dir explicitly.

js731ca replied:
is this a usecase?
`top-dir = /some/place`
`dl-dir = /else/where/downloads`
otherwise one could rename the config to
  `dl-dir-name = downloads`
and it is then always constructed from
  `${top-dir}/${dl-dir-name}`
... why is it configurable in the first place? if its always below top-dir

kanavin replied:
Because you indeed might want to set it /else/where/downloads, e.g. to have a one global dl-dir for everything fetched, layers, configurations, and indeed what recipe put in SRC_URI. But the default is to contain that inside top-dir.

js731ca replied:
so we actually need to remove the line
`builtin_settings['default']['dl-dir'] = os.path.join(top_dir, '.bitbake-setup-downloads')`
as it forces dl-dir to be inside the top_dir - or would reset it.
otherwise a dl-dir outside the top_dir would not be possible :-S

js731ca commented:
f382e7c is just a rebase onto bitbake:master - no changes (yet)

js731ca commented:
reading the bitbake-setup docs that are [going upstream](https://lore.kernel.org/bitbake-devel/20251103-bitbake-setup-v1-2-635a7c829a4b@bootlin.com/):

>    You can set custom new section to ``bitbake-setup`` settings, but the supported setting listed in the :ref:`bitbake-user-manual/bitbake-user-manual-environment-setup:Settings` section are only affected when set in the ``default`` section.

makes me wonder if we shouldnt go back to v1 2a9afbd of this PR, and allow other sections and keys = drop all the checks/enforcements added here

(but keep the change regarding dl-dir=/somewhere/else)

kanavin commented:
> reading the bitbake-setup docs that are [going upstream](https://lore.kernel.org/bitbake-devel/20251103-bitbake-setup-v1-2-635a7c829a4b@bootlin.com/):
> 
> > You can set custom new section to `bitbake-setup` settings, but the supported setting listed in the :ref:`bitbake-user-manual/bitbake-user-manual-environment-setup:Settings` section are only affected when set in the `default` section.
> 
> makes me wonder if we shouldnt go back to v1 [2a9afbd](https://github.com/kanavin/bitbake/commit/2a9afbd407422b8b93ada5b3f21ec464fcb57d57) of this PR, and allow other sections and keys = drop all the checks/enforcements added here
> 
> (but keep the change regarding dl-dir=/somewhere/else)

I agree, let's drop the sanity/enforcements, it's too complex, and brittle, and doesn't seem like worth doing.

js731ca committed:
bitbake-setup: fix key-error when changing a new 'section'

Setting a key=value in a new section would raise a key-error when
using in-memory settings, e.g:

./bin/bitbake-setup --setting default top-dir-prefix /tmp/bitbake-setup --setting default top-dir-name gs settings foo bar baz

Loading settings from
    /tmp/bitbake-setup/gs/settings.conf

Traceback (most recent call last):
  File "/tmp/bitbake-setup/bitbake/./bin/bitbake-setup", line 853, in <module>
    main()
  File "/tmp/bitbake-setup/bitbake/./bin/bitbake-setup", line 838, in main
    all_settings = merge_settings(builtin_settings, global_settings, topdir_settings, args.setting)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/bitbake-setup/bitbake/./bin/bitbake-setup", line 733, in merge_settings
    all_settings[section][setting] = value
    ~~~~~~~~~~~~^^^^^^^^^
KeyError: 'foo'

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca commented:
v4: rebase onto bitbake:mager
( after a slight detour in v3 ccf0bcc00 > 5a8d1f333 > 1b7e78dd9 > 34617eb7a > f382e7cdb )

- going back to v1 ( 2a9afbd ) `fix key-error when changing a new 'section'`
- plus a slight refactoring to use "topdir_settings" more consistently
